{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "query = '/mnt/869990e7-a61f-469f-99fe-a48d24ac44ca/git/ebi/query.fa'\n",
    "target = '/mnt/869990e7-a61f-469f-99fe-a48d24ac44ca/git/ebi/gencode.v46.transcripts.200bp.fa'\n",
    "\n",
    "import sys\n",
    "import os.path as op\n",
    "import ctypes as ct\n",
    "\n",
    "class CAlignRes(ct.Structure):\n",
    "    _fields_ = [('nScore', ct.c_uint16), \n",
    "                ('nScore2', ct.c_uint16), \n",
    "                ('nRefBeg', ct.c_int32), \n",
    "                ('nRefEnd', ct.c_int32), \n",
    "                ('nQryBeg', ct.c_int32), \n",
    "                ('nQryEnd', ct.c_int32), \n",
    "                ('nRefEnd2', ct.c_int32), \n",
    "                ('sCigar', ct.POINTER(ct.c_uint32)), \n",
    "                ('nCigarLen', ct.c_int32)] \n",
    "\n",
    "class CProfile(ct.Structure):\n",
    "    _fields_ = [('pByte', ct.POINTER(ct.c_int32)),\n",
    "                ('pWord', ct.POINTER(ct.c_int32)),\n",
    "                ('pRead', ct.POINTER(ct.c_int8)),\n",
    "                ('pMat', ct.POINTER(ct.c_int8)),\n",
    "                ('nReadLen', ct.c_int32),\n",
    "                ('nN', ct.c_int32),\n",
    "                ('nBias', ct.c_uint8)]\n",
    "\n",
    "class CSsw(object):\n",
    "    def __init__(self, sLibPath):\n",
    "\n",
    "        sLibName = 'libssw.so'\n",
    "        if sLibPath:\n",
    "            if not op.exists(op.join(sLibPath, sLibName)):\n",
    "                sys.stderr.write('libssw.so does not exist in the input path.')\n",
    "                sys.exit()\n",
    "            self.ssw = ct.cdll.LoadLibrary(op.join(sLibPath,sLibName))\n",
    "        else:\n",
    "            bFound = False\n",
    "            for s in sys.path:\n",
    "                if op.exists(op.join(s,sLibName)):\n",
    "                    bFound = True\n",
    "                    self.ssw = ct.cdll.LoadLibrary(op.join(s,sLibName))\n",
    "            if bFound == False:\n",
    "                sys.stderr.write('libssw.so does not exist in PATH')\n",
    "                sys.exit()\n",
    "\n",
    "        self.ssw_init = self.ssw.ssw_init\n",
    "        self.ssw_init.argtypes = [ct.POINTER(ct.c_int8), ct.c_int32, ct.POINTER(ct.c_int8), ct.c_int32, ct.c_int8]\n",
    "        self.ssw_init.restype = ct.POINTER(CProfile)\n",
    "        self.init_destroy = self.ssw.init_destroy\n",
    "        self.init_destroy.argtypes = [ct.POINTER(CProfile)]\n",
    "        self.init_destroy.restype = None\n",
    "        self.ssw_align = self.ssw.ssw_align\n",
    "        self.ssw_align.argtypes = [ct.c_void_p, ct.POINTER(ct.c_int8), ct.c_int32, ct.c_uint8, ct.c_uint8, ct.c_uint8, ct.c_uint16, ct.c_int32, ct.c_int32]\n",
    "        self.ssw_align.restype = ct.POINTER(CAlignRes)\n",
    "        self.align_destroy = self.ssw.align_destroy\n",
    "        self.align_destroy.argtypes = [ct.POINTER(CAlignRes)]\n",
    "        self.align_destroy.restype = None\n",
    "\n",
    "def read_matrix(sFile):\n",
    "    with open(args.sMatrix, 'r') as f:\n",
    "        for l in f:\n",
    "            if not l.startswith('#'):\n",
    "                break\n",
    "        lEle = l.strip().split()\n",
    "        dEle2Int = {}\n",
    "        dInt2Ele = {}\n",
    "        for i,ele in enumerate(lEle):\n",
    "            dEle2Int[ele] = i\n",
    "            dEle2Int[ele.lower()] = i\n",
    "            dInt2Ele[i] = ele\n",
    "        nEleNum = len(lEle)\n",
    "        lScore = []\n",
    "        for l in f:\n",
    "            lScore.extend([int(x) for x in l.strip().split()[1:]])\n",
    "\n",
    "        return lEle, dEle2Int, dInt2Ele, lScore\n",
    "\n",
    "\n",
    "\n",
    "import sys\n",
    "import os.path as op\n",
    "import argparse as ap\n",
    "import ctypes as ct\n",
    "import timeit as ti\n",
    "import gzip\n",
    "import math\n",
    "import ssw_lib\n",
    "\n",
    "\n",
    "def read(sFile):\n",
    "    \"\"\"\n",
    "    read a sequence file\n",
    "    @param  sFile   sequence file\n",
    "    \"\"\"\n",
    "    def read_one_fasta(f):\n",
    "        \"\"\"\n",
    "        read a fasta file\n",
    "        @param  f   file handler\n",
    "        \"\"\"\n",
    "        sId = ''\n",
    "        sSeq = ''\n",
    "        for l in f:\n",
    "            if l.startswith('>'):\n",
    "                if sSeq:\n",
    "                    yield sId, sSeq, ''\n",
    "                sId = l.strip()[1:].split()[0]\n",
    "                sSeq = ''\n",
    "            else:\n",
    "                sSeq += l.strip()\n",
    "\n",
    "        yield sId, sSeq, ''\n",
    "\n",
    "    bFasta = True\n",
    "    with open(sFile, 'r') as f:\n",
    "        l = f.readline()\n",
    "        if l.startswith('>'):\n",
    "            bFasta = True\n",
    "        elif l.startswith('@'):\n",
    "            bFasta = False\n",
    "        else:\n",
    "            sys.stderr.write('file format cannot be recognized\\n')\n",
    "            sys.exit()\n",
    "\n",
    "    with open(sFile, 'r') as f:\n",
    "        if bFasta == True:\n",
    "            for sId,sSeq,sQual in read_one_fasta(f):\n",
    "                yield sId, sSeq, sQual\n",
    "\n",
    "\n",
    "def to_int(seq, lEle, dEle2Int):\n",
    "    \"\"\"\n",
    "    translate a sequence into numbers\n",
    "    @param  seq   a sequence\n",
    "    \"\"\"\n",
    "    num_decl = len(seq) * ct.c_int8\n",
    "    num = num_decl()\n",
    "    for i,ele in enumerate(seq):\n",
    "        try:\n",
    "            n = dEle2Int[ele]\n",
    "        except KeyError:\n",
    "            n = dEle2Int[lEle[-1]]\n",
    "        finally:\n",
    "            num[i] = n\n",
    "\n",
    "    return num\n",
    "\n",
    "\n",
    "def align_one(ssw, qProfile, rNum, nRLen, nOpen, nExt, nFlag, nMaskLen):\n",
    "    \"\"\"\n",
    "    align one pair of sequences\n",
    "    @param  qProfile   query profile\n",
    "    @param  rNum   number array for reference\n",
    "    @param  nRLen   length of reference sequence\n",
    "    @param  nFlag   alignment flag\n",
    "    @param  nMaskLen   mask length\n",
    "    \"\"\"\n",
    "    res = ssw.ssw_align(qProfile, rNum, ct.c_int32(nRLen), nOpen, nExt, nFlag, 0, 0, nMaskLen)\n",
    "\n",
    "    nScore = res.contents.nScore\n",
    "    nScore2 = res.contents.nScore2\n",
    "    nRefBeg = res.contents.nRefBeg\n",
    "    nRefEnd = res.contents.nRefEnd\n",
    "    nQryBeg = res.contents.nQryBeg\n",
    "    nQryEnd = res.contents.nQryEnd\n",
    "    nRefEnd2 = res.contents.nRefEnd2\n",
    "    lCigar = [res.contents.sCigar[idx] for idx in range(res.contents.nCigarLen)]\n",
    "    nCigarLen = res.contents.nCigarLen\n",
    "    ssw.align_destroy(res)\n",
    "\n",
    "    return (nScore, nScore2, nRefBeg, nRefEnd, nQryBeg, nQryEnd, nRefEnd2, nCigarLen, lCigar)\n",
    "\n",
    "\n",
    "def buildPath(q, r, nQryBeg, nRefBeg, lCigar):\n",
    "    \"\"\"\n",
    "    build cigar string and align path based on cigar array returned by ssw_align\n",
    "    @param  q   query sequence\n",
    "    @param  r   reference sequence\n",
    "    @param  nQryBeg   begin position of query sequence\n",
    "    @param  nRefBeg   begin position of reference sequence\n",
    "    @param  lCigar   cigar array\n",
    "    \"\"\"\n",
    "    sCigarInfo = 'MIDNSHP=X'\n",
    "    sCigar = ''\n",
    "    sQ = ''\n",
    "    sA = ''\n",
    "    sR = ''\n",
    "    nQOff = nQryBeg\n",
    "    nROff = nRefBeg\n",
    "    for x in lCigar:\n",
    "        n = x >> 4\n",
    "        m = x & 15\n",
    "        if m > 8:\n",
    "            c = 'M'\n",
    "        else:\n",
    "            c = sCigarInfo[m]\n",
    "        sCigar += str(n) + c\n",
    "\n",
    "        if c == 'M':\n",
    "            sQ += q[nQOff : nQOff+n]\n",
    "            sA += ''.join(['|' if q[nQOff+j] == r[nROff+j] else '*' for j in range(n)])\n",
    "            sR += r[nROff : nROff+n]\n",
    "            nQOff += n\n",
    "            nROff += n\n",
    "        elif c == 'I':\n",
    "            sQ += q[nQOff : nQOff+n]\n",
    "            sA += ' ' * n\n",
    "            sR += '-' * n\n",
    "            nQOff += n\n",
    "        elif c == 'D':\n",
    "            sQ += '-' * n\n",
    "            sA += ' ' * n\n",
    "            sR += r[nROff : nROff+n]\n",
    "            nROff += n\n",
    "    return sCigar, sQ, sA, sR\n",
    "\n",
    "def main():\n",
    "    nMatch = 2\n",
    "    nMismatch = 2\n",
    "\n",
    "    nOpen = 3\n",
    "    nExt = 1\n",
    "\n",
    "    lEle = ['A', 'C', 'G', 'T', 'N']\n",
    "    dRc = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'a': 'T', 'c': 'G', 'g': 'C', 't': 'A'}\n",
    "    dEle2Int = {}\n",
    "    dInt2Ele = {}\n",
    "\n",
    "    for i, ele in enumerate(lEle):\n",
    "        dEle2Int[ele] = i\n",
    "        dEle2Int[ele.lower()] = i\n",
    "        dInt2Ele[i] = ele\n",
    "\n",
    "    nEleNum = len(lEle)\n",
    "    lScore = [0 for _ in range(nEleNum ** 2)]\n",
    "    for i in range(nEleNum - 1):\n",
    "        for j in range(nEleNum - 1):\n",
    "            if lEle[i] == lEle[j]:\n",
    "                lScore[i * nEleNum + j] = nMatch\n",
    "            else:\n",
    "                lScore[i * nEleNum + j] = -nMismatch\n",
    "\n",
    "    mat = (len(lScore) * ct.c_int8)()\n",
    "    mat[:] = lScore\n",
    "\n",
    "    nFlag = 0\n",
    "    ssw = ssw_lib.CSsw('./')\n",
    "\n",
    "    for sQId, sQSeq, sQQual in read(query):\n",
    "        qNum = to_int(sQSeq, lEle, dEle2Int)\n",
    "        qProfile = ssw.ssw_init(qNum, ct.c_int32(len(sQSeq)), mat, len(lEle), 2)\n",
    "        nMaskLen = len(sQSeq) // 2\n",
    "\n",
    "        best_score = -float('inf')\n",
    "        best_alignment = None\n",
    "        ref = None\n",
    "\n",
    "        for sRId, sRSeq, _ in read(target):\n",
    "            rNum = to_int(sRSeq, lEle, dEle2Int)\n",
    "            res = align_one(ssw, qProfile, rNum, len(sRSeq), nOpen, nExt, nFlag, nMaskLen)\n",
    "\n",
    "            if res[0] > best_score:\n",
    "                ref = sRId\n",
    "                best_score = res[0]\n",
    "                best_alignment = res\n",
    "\n",
    "        if best_alignment is not None:\n",
    "            print(f\"Query ID: {sQId}\")\n",
    "            print(f\"Ref ID: {ref}\")\n",
    "            print(f\"Best Score: {best_score}\")\n",
    "            print(f\"Best : {res}\")\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "254065\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "counter = 0\n",
    "\n",
    "input_fasta = \"/mnt/869990e7-a61f-469f-99fe-a48d24ac44ca/git/ebi/gencode.v46.transcripts.fa\"\n",
    "output_fasta = \"/mnt/869990e7-a61f-469f-99fe-a48d24ac44ca/git/ebi/gencode.v46.transcripts.short.fa\"\n",
    "# def extract_last_200_bp(input_file, output_file):\n",
    "with open(output_fasta, \"w\") as out_handle:\n",
    "    for record in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "        if len(record.seq) > 15:\n",
    "            counter = counter + 1\n",
    "            # Extract the last 200 bp of the sequence\n",
    "            last_200_bp = record.seq[-100:]\n",
    "            # Create a new record with the truncated sequence\n",
    "            new_record = record[:]\n",
    "            new_record.seq = last_200_bp\n",
    "            # Write the new record to the output file\n",
    "            SeqIO.write(new_record, out_handle, \"fasta\")\n",
    "\n",
    "# Replace these paths with your actual file paths\n",
    "\n",
    "\n",
    "# extract_last_200_bp(input_fasta, output_fasta)\n",
    "\n",
    "print(counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import random\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def extract_last_bases(seq, length=20):\n",
    "    \"\"\"Extract the last 'length' bases from a sequence.\"\"\"\n",
    "    return seq[-length:]\n",
    "\n",
    "def process_fasta(input_fasta, output_fasta, min_length=15, num_sequences=512, last_bases_length=50):\n",
    "    # Read all records from the input FASTA file\n",
    "    records = [record for record in SeqIO.parse(input_fasta, \"fasta\") if len(record.seq) > min_length]\n",
    "    \n",
    "    # Check if there are enough records to sample\n",
    "    if len(records) < num_sequences:\n",
    "        raise ValueError(f\"Not enough sequences longer than {min_length} bp in {input_fasta} to sample {num_sequences} sequences.\")\n",
    "    \n",
    "    # Randomly select the specified number of sequences\n",
    "    random.seed(123)\n",
    "    selected_records = random.sample(records, num_sequences)\n",
    "    \n",
    "    # Extract the last 'last_bases_length' bases from each selected sequence\n",
    "    last_bases_records = []\n",
    "    for record in selected_records:\n",
    "        last_bases_seq = extract_last_bases(record.seq, last_bases_length)\n",
    "        last_bases_record = SeqRecord(last_bases_seq, id=record.id, description=record.description)\n",
    "        last_bases_records.append(last_bases_record)\n",
    "    \n",
    "    # Write the extracted sequences to the output FASTA file\n",
    "    with open(output_fasta, \"w\") as out_handle:\n",
    "        SeqIO.write(last_bases_records, out_handle, \"fasta\")\n",
    "\n",
    "# Example usage\n",
    "input_fasta = \"/mnt/869990e7-a61f-469f-99fe-a48d24ac44ca/git/ebi/gencode.v46.transcripts.fa\"\n",
    "output_fasta = \"/mnt/869990e7-a61f-469f-99fe-a48d24ac44ca/git/ebi/gencode.v46.transcripts.50bp.fa\"\n",
    "process_fasta(input_fasta, output_fasta)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
